src/dbnode/persist/fs/seek.go

```diff
@@ -383,55 +368,96 @@ func (s *seeker) SeekByIndexEntry(entry IndexEntry) (checked.Bytes, error) {
 	return buffer, nil
 }
 
-func (s *seeker) SeekIndexEntry(id ident.ID) (IndexEntry, error) {
-	offset, err := s.indexLookup.getNearestIndexFileOffset(id)
+func (s *seeker) SeekIndexEntry(
+	id ident.ID,
+	resources ReusableSeekerResources,
+) (IndexEntry, error) {
+	offset, err := s.indexLookup.getNearestIndexFileOffset(id, resources)
 	// Should never happen, either something is really wrong with the code or
-	// the file on disk was corrupted
+	// the file on disk was corrupted.
+	if err != nil {
+		return IndexEntry{}, err
+	}
+
+	seekedOffset, err := s.indexFd.Seek(offset, os.SEEK_SET)
 	if err != nil {
 		return IndexEntry{}, err
 	}
+	if seekedOffset != offset {
+		return IndexEntry{}, fmt.Errorf("tried to seek to offset: %d, but seeked to: %d", seekedOffset, offset)
+	}
 
-	stream := msgpack.NewDecoderStream(s.indexMmap[offset:])
-	s.decoder.Reset(stream)
+	resources.fileDecoderStream.Reset(s.indexFd)
+	resources.xmsgpackDecoder.Reset(resources.fileDecoderStream)
 
 	idBytes := id.Bytes()
-	// Prevent panic's when we're scanning to the end of the buffer
-	for stream.Remaining() != 0 {
-		entry, err := s.decoder.DecodeIndexEntry()
-		// Should never happen, either something is really wrong with the code or
-		// the file on disk was corrupted
+	for {
+		currOffset, err := s.indexFd.Seek(0, os.SEEK_CUR)
```

Comment:
So like 395 is not actually changing the offset, thats actually just how you ask the fd to tell you what the current offset is.

Basically what happens is we:

```
1. Go to the indexLookup thing and it will give us an offset that is a good starting point for scanning
2. Line 382 we seek to the position that the indexLookup told us would be a good start
3. We reset a decoder with our fileDecoderStream (basically a fd wrapped in a bufio.reader)
4. We call entry.DecodeIndexEntry() in a tight loop until we've either: found the entry, gone too far, or reached the end of a file. The further seeks are handled automatically by the fileDecoderStream/bufioReader/fd
```

Reply:
Do you want me to add a comment or something with basically what I just wrote above?

Reply:
I just added a comment to it